src/whitelisting.ts

```diff
@@ -0,0 +1,42 @@
+import { NextFunction, Request, Response } from 'express'
+import Router from 'express-promise-router'
+
+import { getStoreSettings } from './storeSettings'
+import { makeApiClientError } from './utils'
+
+export async function whitelistIps(
+  req: Request,
+  _res: Response,
+  next: NextFunction
+): Promise<void> {
+  const { ipWhitelist } = await getStoreSettings()
+
+  const clientIp = req.ip
+
+  if (ipWhitelist.length > 0 && !ipWhitelist.includes(clientIp)) {
+    next(makeApiClientError(403, 'Forbidden IP'))
+  }
+
+  next()
+}
+
+export async function whitelistApiKeys(
+  req: Request,
+  _res: Response,
+  next: NextFunction
+): Promise<void> {
+  const { apiKeyWhitelist } = await getStoreSettings()
+
+  const clientApiKey = req.query.apiKey
+
+  if (apiKeyWhitelist.length > 0 && !apiKeyWhitelist.includes(clientApiKey)) {
+    next(makeApiClientError(403, 'Forbidden API Key'))
+  }
+
+  next()
+}
+
+export const whitelistAll = Router()
```

Comment:
will express return a failure if only one of the conditions succeeds? The server should allow either a valid API key or a valid IP address. I just want to make sure that we don’t fail the call if one passes and the other does not

Reply:
If the condition is an OR, then I'll change this router composition.